src/rgw/rgw_admin.cc

```diff
@@ -2181,6 +2195,32 @@ static void get_data_sync_status(const string& source_zone, list<string>& status
         push_ss(ss, status, tab) << "oldest incremental change not applied: " << oldest;
       }
     }
+  } 
+
+  if (error_shards.size() > 0) {
+    push_ss(ss, status, tab) << error_shards.size() << " shards have encountered errors";
+    push_ss(ss, status, tab) << "error shards: " << "[" << error_shards << "]";
+  }
+
+  if (sync_detail) {
+    ret = sync.read_lagging_buckets(&sync_status, shards_behind, error_shards); // read error shards
```

Comment:
i'm not sure how i feel about this 'lagging buckets' part

on the one hand, it's useful to see which buckets could be affected by the datalog shards that are behind. but on the other hand, it's the bucket's sync status that tells whether the bucket is actually behind. so there will be many cases where buckets have finished sync but are reported as laggy here because there are still some datalog entries referring to them (this would often be due to failures in other buckets on the same datalog shard that delay progress). so i worry that false positives here will confuse the admin, and they'll assume that they need to do something to fix all of those buckets

it's also fairly expensive to compute - first, `RGWReadRemoteDataLogCR` will read up to 1000 datalog entries from each shard (so up to 128000 entries in memory), then `RGWReadDataSyncErrorShardsCR` reads all of the error entries (which is not bounded in size). in large deployments, this is likely to either crash radosgw-admin, or just take a really long time and produce too much output

Reply:
thank you for kindly review. maybe just use bucket's sync status can not get all of the lagging bucket shards(i.e the related bucket instance have't been synced), instead of this ```lagging buckets``` part,  maybe we can do as below?
1. get the behind data log shards and error data log shards
2. for every behind data log shard, read the related bucket instance, then use bucket's sync status to judge whether the bucket is actually behind
3. for every error data log shard, just read the error bucket shards record in the related ```*.retry obj```

any suggestions are appreciated!

Reply:
I agree with @cbodley here. I think the proper way to do it would be to flip the responsibility. Instead of having `radosgw-admin` querying the status of buckets, have `radosgw` report the status somewhere durable when the state changes. It is already being done for the sync trace node info, just that we currently only keep it in memory.

Reply:
@Leeshine i would really like to see improvements to the `bucket sync status` command so it can query the remote zones and compare their markers with the local bucket sync status. i think that would be the ideal way to inspect a single bucket/shard

so maybe the overall approach should be to start with a `sync status` command that tells which datalog shards are behind (as you've done). then have some kind of `data sync status --shard=x` command that lists which bucket shards are left to process (this 'lagging buckets' part). then the admin could use `bucket sync status` to look into the status of each one

that way we don't make the `sync status` command itself too complicated or verbose, but provide more tools to inspect the status in detail. what do you guys think?

Reply:
@cbodley sounds good!

Reply:
cc @vumrao

Reply:
all right, I'll try to working on it. @cbodley

Reply:
@Leeshine can we limit this pr to just the `sync status` changes (minus --sync-detail and 'lagging buckets') and follow up with the other commands separately? i think what you've done so far is close to being ready to merge

Reply:
@cbodley  sure